google/cloud/pubsub_v1/publisher/client.py

```diff
@@ -133,6 +133,12 @@ def __init__(self, batch_settings=(), publisher_options=(), **kwargs):
                 target=os.environ.get("PUBSUB_EMULATOR_HOST")
             )
 
+        client_options = kwargs.pop("client_options", None)
+        if client_options and type(client_options["api_endpoint"]) is str:
```

Comment:
Two things:
- If `api_endpoint` is a unicode string under Python 2, the type check will be negative and the option will (still) be ignored:
    ```py
    >>> type(u"foo/bar.com") is str
    False
    ```
    It is thus better to rewrite it as `isinstance(maybe_string, six.string_types)` (it does return `False` for `bytes` under Python 3, but this is fine, as the latter makes a clear distinction between text and binary data).
- if `client_options` does not contain the `"api_endpoint"` key, a `KeyError` will be raised, which is probably undesired, and we should just fall back to the default SERVICE_ADDRESS.

Let's thus rewrite this snippet, and add a test or two for the unhappy path, too. Applies to the subscriber client below, too.

Reply:
Thanks, done!